File: Java/Lesson084_KthSmallestElementInABst.java
// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 084 -- Kth Smallest Element in a BST
// Category   : Trees
// Difficulty : Medium
// Study Plan : Day 42
// =============================================================
//
// QUESTION:
//   Given a BST, return the kth smallest value (1-indexed). Use inorder traversal.
// =============================================================

import java.util.*;
public class Lesson084_KthSmallestElementInABst {
    static class TreeNode { int val; TreeNode left, right; TreeNode(int v, TreeNode l, TreeNode r){val=v; left=l; right=r;} TreeNode(int v){this(v,null,null);} }
    public int kthSmallest(TreeNode root, int k){
        Deque<TreeNode> st = new ArrayDeque<>(); TreeNode cur = root;
        while (!st.isEmpty() || cur != null){
            while (cur != null){ st.push(cur); cur = cur.left; }
            cur = st.pop(); if (--k == 0) return cur.val;
            cur = cur.right;
        }
        return -1;
    }
    public static void main(String[] a){
        TreeNode r = new TreeNode(3, new TreeNode(1, null, new TreeNode(2)), new TreeNode(4));
        System.out.println(new Lesson084_KthSmallestElementInABst().kthSmallest(r, 1)+" "+new Lesson084_KthSmallestElementInABst().kthSmallest(r, 3));
    }
}

File: Go/Lesson_084_kth_smallest_element_in_a_bst.go
//go:build ignore

// =============================================================
// MIT License | @analyticswithharry2026
// GitHub  : https://github.com/analyticswithharry
// YouTube : Analytics with Harry
// =============================================================
// Lesson     : 084 -- Kth Smallest Element in a BST
// Category   : Trees
// Difficulty : Medium
// Study Plan : Day 42
// =============================================================
//
// QUESTION:
//   Given a BST, return the kth smallest value (1-indexed). Use inorder traversal.
// =============================================================

package main
import "fmt"
type TreeNode struct { Val int; Left, Right *TreeNode }
func kthSmallest(root *TreeNode, k int) int {
    st := []*TreeNode{}; cur := root
    for len(st) > 0 || cur != nil {
        for cur != nil { st = append(st, cur); cur = cur.Left }
        cur = st[len(st)-1]; st = st[:len(st)-1]; k--
        if k == 0 { return cur.Val }
        cur = cur.Right
    }
    return -1
}
func main(){
    r := &TreeNode{Val:3, Left:&TreeNode{Val:1, Right:&TreeNode{Val:2}}, Right:&TreeNode{Val:4}}
    fmt.Println(kthSmallest(r,1), kthSmallest(r,3))
}
